### DISCUSSION TITLE
remove zero delta tuples from variable font

### DISCUSSION THREAD

**Discussion by aminabedi68** [fonttools/fonttools#2830] [Q&A] [answered]
Hi
in many projects and most of the time, a large number of glyphs remain unchanged along the axes, this cause to generate a large number of unnecessary zero delta tuples, like this:
```
      <tuple>
        <coord axis="wght" min="0.375" value="1.0" max="1.0"/>
        <coord axis="CNTR" value="1.0"/>
        <delta pt="0" x="0" y="0"/>
        <delta pt="1" x="0" y="0"/>
        <delta pt="2" x="0" y="0"/>
        <delta pt="3" x="0" y="0"/>
        <delta pt="4" x="0" y="0"/>
        <delta pt="5" x="0" y="0"/>
        <delta pt="6" x="0" y="0"/>
      </tuple>
```
removing these tuples helps to optimize the file size. glyphs app has `Optimize Variable Delta` for this, [Slice ](https://github.com/source-foundry/Slice)also do similar optimization...

I need it as a part of independent build process. maybe it's predicted in fonttools or ttx, or a script(that i couldn't write it). so please help.
thank you.

(I'm not aware of software complexities, but extremely recommend to add it (directly or as an optional stage) to `fontmake`.)

**Comment by behdad** [comment, marked as answer]
Something like this should do:
```python
from fontTools.ttLib import TTFont
import sys

font = TTFont(sys.argv[1])
gvar = font["gvar"]

new_variations = {}
for glyph,variations in gvar.variations.items():
    l1 = len(variations)
    variations = [v for v in variations
                  if any(i is None or i[0] or i[1]
                         for i in v.coordinates)]
    l2 = len(variations)
    if l1 != l2:
        print(glyph, l1, l2)
    if variations:
        new_variations[glyph] = variations
gvar.variations = new_variations

font.save("out.ttf")
```

**Reply by aminabedi68** [reply]
thank you @behdad
no, not works for me.

**Reply by behdad** [reply]
> no, not works for me.

Send me your font, I'll take a look.

**Reply by behdad** [reply]
Running this script on RobotoFlex saves about 13% of the font size. @davelab6

**Reply by aminabedi68** [reply]
[Mikhak typeface](https://github.com/aminabedi68/Mikhak/blob/master/fonts/Variable/Mikhak%5BDSTY%2CKSHD%2Cwght%5D.ttf)

**Reply by behdad** [reply]
> [Mikhak typeface](https://github.com/aminabedi68/Mikhak/blob/master/fonts/Variable/Mikhak%5BDSTY%2CKSHD%2Cwght%5D.ttf)

And out of curiosity, what toolchain built this font?

**Reply by behdad** [reply]
The script seems to work on your font. Why do you think it doesn't? Just add a `font.save("out.ttf")` at the end. For me:
```
-rw-r--r--. 1 behdad behdad 197852 Sep 21 13:46 'Mikhak[DSTY,KSHD,wght].ttf'
-rw-r--r--. 1 behdad behdad 140724 Sep 21 13:47  out.ttf
```

**Reply by behdad** [reply]
Wait. It looks like all of those correspond to composite glyphs. We encode those for a reason. There was a bug in CoreText I think...

If I change the snippet to only do the optimization for non-composite glyphs, then your font will be unchanged, suggesting that it was built with fontmake. Is that correct

**Reply by aminabedi68** [reply]
> > [Mikhak typeface](https://github.com/aminabedi68/Mikhak/blob/master/fonts/Variable/Mikhak%5BDSTY%2CKSHD%2Cwght%5D.ttf)
> 
> And out of curiosity, what toolchain built this font?

fontforge, fonttools, fontmake,...



> The script seems to work on your font. Why do you think it doesn't? Just add a `font.save("out.ttf")` at the end. For me:
> 
> ```
> -rw-r--r--. 1 behdad behdad 197852 Sep 21 13:46 'Mikhak[DSTY,KSHD,wght].ttf'
> -rw-r--r--. 1 behdad behdad 140724 Sep 21 13:47  out.ttf
> ```

i know, i added the line and ran a diff between, the output is the same.

**Reply by aminabedi68** [reply]
> Wait. It looks like all of those correspond to composite glyphs. We encode those for a reason. There was a bug in CoreText I think...
> 
> If I change the snippet to only do the optimization for non-composite glyphs, then your font will be unchanged, suggesting that it was built with fontmake. Is that correct

yes.

**Reply by aminabedi68** [reply]
i tried again with your latest edition, it's working now :)
should i use it? i mean with the CoreText  bug that you mentioned?

**Reply by behdad** [reply]
I'm coming up with a better fix in https://github.com/fonttools/fonttools/issues/2831 which has the CoreText workaround but is more optimized.

**Reply by behdad** [reply]
Can you build your font with fonttools from https://github.com/fonttools/fonttools/pull/2832 ?

**Reply by aminabedi68** [reply]
I'm sorry. Unfortunately I couldn't do it. I'm not familiar with this much about github.

**Comment by behdad** [comment]
> (I'm not aware of software complexities, but extremely recommend to add it (directly or as an optional stage) to fontmake.)

We already do that:
https://github.com/fonttools/fonttools/blob/main/Lib/fontTools/varLib/__init__.py#L275

except for composite glyphs as I mentioned.  The change that disabled this optimization for composites was:
https://github.com/fonttools/fonttools/commit/9c3dde3ce6be74b7d06127c2d9a0c17484d638a0

**Reply by aminabedi68** [reply]
ok, i didn't aware of that, thank you.
